src/posix/platform/multicast_routing.cpp

```diff
@@ -0,0 +1,569 @@
+{
+    bool                                      found = false;
+    otBackboneRouterMulticastListenerIterator iter  = OT_BACKBONE_ROUTER_MULTICAST_LISTENER_ITERATOR_INIT;
+    otBackboneRouterMulticastListenerInfo     listenerInfo;
+
+    while (otBackboneRouterMulticastListenerGetNext(mInstance, &iter, &listenerInfo) == OT_ERROR_NONE)
+    {
+        VerifyOrExit(static_cast<const Ip6::Address &>(listenerInfo.mAddress) != aAddress, found = true);
+    }
+
+exit:
+    return found;
+}
+
+void MulticastRoutingManager::UpdateFdSet(fd_set &aReadFdSet, int &aMaxFd) const
+{
+    VerifyOrExit(IsEnabled());
+
+    FD_SET(mMulticastRouterSock, &aReadFdSet);
+    aMaxFd = OT_MAX(aMaxFd, mMulticastRouterSock);
+
+exit:
+    return;
+}
+
+void MulticastRoutingManager::Process(const fd_set &aReadFdSet)
+{
+    VerifyOrExit(IsEnabled());
+
+    ExpireMulticastForwardingCache();
+
+    if (FD_ISSET(mMulticastRouterSock, &aReadFdSet))
+    {
+        ProcessMulticastRouterMessages();
+    }
+
+exit:
+    return;
+}
+
+void MulticastRoutingManager::InitMulticastRouterSock(void)
+{
+    int                 one = 1;
+    struct icmp6_filter filter;
+    struct mif6ctl      mif6ctl;
+
+    // Create a Multicast Routing socket
+    mMulticastRouterSock = socket(AF_INET6, SOCK_RAW, IPPROTO_ICMPV6);
+    VerifyOrDie(mMulticastRouterSock != -1, OT_EXIT_ERROR_ERRNO);
+
+    // Enable Multicast Forwarding in Kernel
+    VerifyOrDie(0 == setsockopt(mMulticastRouterSock, IPPROTO_IPV6, MRT6_INIT, &one, sizeof(one)), OT_EXIT_ERROR_ERRNO);
+
+    // Filter all ICMPv6 messages
+    ICMP6_FILTER_SETBLOCKALL(&filter);
+    VerifyOrDie(0 == setsockopt(mMulticastRouterSock, IPPROTO_ICMPV6, ICMP6_FILTER, (void *)&filter, sizeof(filter)),
+                OT_EXIT_ERROR_ERRNO);
+
+    memset(&mif6ctl, 0, sizeof(mif6ctl));
+    mif6ctl.mif6c_flags     = 0;
+    mif6ctl.vifc_threshold  = 1;
+    mif6ctl.vifc_rate_limit = 0;
+
+    // Add Thread network interface to MIF
+    mif6ctl.mif6c_mifi = kMifIndexThread;
+    mif6ctl.mif6c_pifi = if_nametoindex(gNetifName);
+    VerifyOrDie(mif6ctl.mif6c_pifi > 0, OT_EXIT_ERROR_ERRNO);
+    VerifyOrDie(0 == setsockopt(mMulticastRouterSock, IPPROTO_IPV6, MRT6_ADD_MIF, &mif6ctl, sizeof(mif6ctl)),
+                OT_EXIT_ERROR_ERRNO);
+
+    // Add Backbone network interface to MIF
+    mif6ctl.mif6c_mifi = kMifIndexBackbone;
+    mif6ctl.mif6c_pifi = if_nametoindex(gBackboneNetifName);
+    VerifyOrDie(mif6ctl.mif6c_pifi > 0, OT_EXIT_ERROR_ERRNO);
+    VerifyOrDie(0 == setsockopt(mMulticastRouterSock, IPPROTO_IPV6, MRT6_ADD_MIF, &mif6ctl, sizeof(mif6ctl)),
+                OT_EXIT_ERROR_ERRNO);
+}
+
+void MulticastRoutingManager::FinalizeMulticastRouterSock(void)
+{
+    VerifyOrExit(IsEnabled());
+
+    close(mMulticastRouterSock);
+    mMulticastRouterSock = -1;
+
+exit:
+    return;
+}
+
+void MulticastRoutingManager::ProcessMulticastRouterMessages(void)
+{
+    otError         error = OT_ERROR_NONE;
+    char            buf[128];
```

Comment:
Is it possible to use sizeof something here?

Reply:
Changed to use `sizeof(struct mrt6msg)`.